### ISSUE TITLE
[Question/Wishlist] Using JSON for the protocol instead of XML

### ISSUE THREAD

**Issue by fwcd** [software-challenge/backend#408]
This is more of a general question rather than a specific issue. Has using JSON instead of XML for the protocol ever been considered, apart from legacy reasons?

While such a change would have quite a big impact on the way the different clients and the server handles communication, I think there are strong reasons for at least considering such a change e.g. for the next year:

* JSON is ubiquitous, simple and easy to parse/serialize in almost every mainstream language (even hand-rolling a parser isn't too hard)
* JSON maps very naturally to data structures in almost every mainstream language (in contrast to XML, where especially serialization can be very ambiguous, with regard to whether e.g. something should be serialized as text, an element, an attribute, etc.)
* JSON is human-readable, debuggable and quite a bit less verbose than XML

A particular pain point with XML that I've encountered myself is e.g. the lack of good library support for Rust, leaving us essentially with either the option of [hand-writing the XML mappings](https://github.com/fwcd/socha-sdk-rust/blob/02a3947df95a0c0f30387477b06dadee566dc8c3/plugin-2021/src/game/game_state.rs#L317-L337) (which is tedious and error-prone) or using pretty limited tools for defining automatic mappings between structs and XML, which still require [careful manual review of certain edge cases](https://github.com/fwcd/socha-sdk-rust/blob/318aa21dbdf03b793af8d77f98be62265b0c73d8/client-base/src/protocol/room.rs#L26-L56).

While these problems might not apply (as much) to other languages, JSON would solve pretty much all of these issues and at the same time simplify the protocol in general.

**Comment by xeruf** [comment]
Yeah, sometimes I have been unhappy with XML, too - but it is pretty stable now and every major language should support it. I am pretty surprised there is no parser for Rust yet.

We have quite a few big construction zones that concern the user-friendliness and maintainability of the project, so if this is ever considered, not in the near-term, sorry.
